### PR TITLE
Paged secure memory

### PR DESCRIPTION
**Pull request by jenswi-linaro**

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
Checkpatch complains about the function pointers (as usual).

File: core/arch/arm/mm/mobj.c [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+struct mobj_phys {
+	TAILQ_ENTRY(mobj_phys) link;
+	struct mobj mobj;
+	enum buf_is_attr battr;
+	uint32_t cattr; /* Defined by TEE_MATTR_CACHE_* in tee_mmu_types.h */
+	vaddr_t va;
+	paddr_t pa;
+};
+
+static TAILQ_HEAD(, mobj_phys) moph_head = TAILQ_HEAD_INITIALIZER(moph_head);
```
Comment (lorc) [inline core/arch/arm/mm/mobj.c]: I can't find where linked list is used.

File: core/arch/arm/mm/mobj.c [hunk 2]
```diff
@@ -294,7 +296,7 @@ struct mobj *mobj_mm_alloc(struct mobj *mobj_parent, size_t size,
 
 static void mobj_paged_free(struct mobj *mobj);
 
-static const struct mobj_ops mobj_paged_ops = {
+static const struct mobj_ops mobj_paged_ops __rodata_unpaged = {
```
Comment (lorc) [inline core/arch/arm/mm/mobj.c]: I think this should be moved to original commit.

File: core/arch/arm/mm/mobj.c (hunk 1 above)
Comment (jenswi-linaro) [inline core/arch/arm/mm/mobj.c]: Sorry, some leftover. I'll remove it.

**Comment by jenswi-linaro** [comment]
Addressed comments

File: core/arch/arm/include/mm/mobj.h [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+	return false;
+}
+
+struct mobj *mobj_mm_alloc(struct mobj *mobj_parent, size_t size,
+			   tee_mm_pool_t *pool);
+
+struct mobj *mobj_phys_alloc(paddr_t pa, size_t size, uint32_t cattr,
+			     enum buf_is_attr battr);
+
+struct mobj *mobj_paged_alloc(size_t size);
```
Comment (lorc) [inline core/arch/arm/include/mm/mobj.h]: In `.h` file `mobj_paged_alloc` is outside `#ifdef CFG_PAGED_USER_TA`. But in `.c` file it is inside that `#ifdef`

File: core/arch/arm/mm/mobj.c [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+	.free = mobj_paged_free,
+};
+
+static void mobj_paged_free(struct mobj *mobj)
+{
+	assert(mobj->ops == &mobj_paged_ops);
+	free(mobj);
+}
+
+struct mobj *mobj_paged_alloc(size_t size)
```
Comment (lorc) [inline core/arch/arm/mm/mobj.c]: In `.c` file `mobj_paged_alloc` is inside `#ifdef CFG_PAGED_USER_TA`. But in `.h` file it is outside that `#ifdef`

File: core/arch/arm/include/mm/mobj.h (hunk 3 above)
Comment (jenswi-linaro) [inline core/arch/arm/include/mm/mobj.h]: This is something we haven't paid any attention to in the past. It clutters more than it helps IMO. I don't think they care much about this in the Linux kernel either.

File: core/arch/arm/mm/pgt_cache.c [hunk 5]
```diff
@@ -340,6 +341,128 @@ void pgt_flush_ctx(struct tee_ta_ctx *ctx)
...
+	mutex_unlock(&pgt_mu);
+}
+
+void pgt_transfer(struct pgt_cache *pgt_cache, void *old_ctx, vaddr_t old_va,
+		  void *new_ctx, vaddr_t new_va, size_t size)
+{
+	const size_t pgtsize = CORE_MMU_PGDIR_SIZE;
+	const vaddr_t new_base = ROUNDDOWN(new_va, pgtsize);
+	const vaddr_t old_base = ROUNDDOWN(old_va, pgtsize);
+	const size_t num_new_pgt = (size - 1 + new_va - new_base) / pgtsize + 1;
```
Comment (lorc) [inline core/arch/arm/mm/pgt_cache.c]: if `size == 0`,  then `num_new_pgt` will be some big value (check http://ideone.com/XID9h3)

File: core/arch/arm/mm/mobj.c [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+{
+	return mobj && mobj->ops == &mobj_sec_shm_ops;
+}
+
+struct mobj *mobj_sec_shm_alloc(size_t size)
+{
+	struct thread_specific_data *tsd = thread_get_tsd();
+	struct mobj_sec_shm *m;
+	struct user_ta_ctx *utc;
+	vaddr_t va;
```
Comment (jenswi-linaro) [inline core/arch/arm/mm/mobj.c]: Thanks, well spotted.

File: core/arch/arm/mm/pgt_cache.c [hunk 7]
```diff
@@ -340,6 +341,128 @@ void pgt_flush_ctx(struct tee_ta_ctx *ctx)
...
+}
+
+static bool pgt_entry_matches(struct pgt *p, void *ctx, vaddr_t begin,
+			      vaddr_t last)
+{
+	if (!p)
+		return false;
+	if (p->ctx != ctx)
+		return false;
+	if (!core_is_buffer_inside(p->vabase, SMALL_PAGE_SIZE, begin,
```
Comment (jenswi-linaro) [inline core/arch/arm/mm/pgt_cache.c]: It "can't happen" in the current code, but you're right.

File: core/arch/arm/mm/pgt_cache.c (hunk 5 above)
Comment (jenswi-linaro) [inline core/arch/arm/mm/pgt_cache.c]: Thanks, I'll fix.

**Comment by jenswi-linaro** [comment]
No, I haven't checked with coverity.

**Comment by jenswi-linaro** [comment]
This PR is replaced by https://github.com/OP-TEE/optee_os/pull/1253 which is rebased on latest and has all fixes so far squashed into the correct commit.

**Pull request closed without merging**